onchain/danaswap/DanaSwap.hs

```diff
@@ -0,0 +1,57 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module DanaSwap (
+  trivialCbor,
+  trivialFailCbor,
+  configScriptCbor,
+  nftCbor,
+) where
+
+import Plutarch.Prelude
+
+import Utils (closedTermToHexString, globalConfig, validatorToHexString)
+
+import Plutarch.Api.V2 (PMintingPolicy, PValidator, mkValidator)
+import Plutarch.Extensions.Api (passert)
+
+import Plutarch.Api.V2.Tx (PTxOutRef)
+import Plutarch.Extensions.Data (parseData)
+
+data HelloRedemer (s :: S)
+  = Inc (Term s (PDataRecord '[]))
+  | Spend (Term s (PDataRecord '[]))
+  deriving stock (Generic)
+  deriving anyclass (PlutusType, PIsData, PEq)
+
+instance DerivePlutusType HelloRedemer where type DPTStrat _ = PlutusTypeData
+
+trivialCbor :: Maybe String
+trivialCbor = closedTermToHexString trivial
+
+trivial :: ClosedTerm (PData :--> PValidator)
+trivial = plam $ \_ _ _ _ -> popaque $ pcon PUnit
+
+trivialFailCbor :: Maybe String
+trivialFailCbor = closedTermToHexString trivialFail
+
+trivialFail :: ClosedTerm PValidator
+trivialFail = perror
+
+configScriptCbor :: String
+configScriptCbor = validatorToHexString $ mkValidator globalConfig configScript
+
+configScript :: ClosedTerm PValidator
+configScript = perror
+
+nftCbor :: Maybe String
+nftCbor = closedTermToHexString standardNft
+
+standardNft :: ClosedTerm (PData :--> PMintingPolicy)
+standardNft = phoistAcyclic $
+  plam $ \outRefData _ sc -> unTermCont $ do
+    outRef :: Term _ PTxOutRef <- parseData outRefData
+    let (inputs :: Term _ (PBuiltinList PTxOutRef)) =
+          pmap # pfield @"outRef"
+            #$ pfield @"inputs"
+            #$ pfield @"txInfo" # sc
+    passert "didn't spend out ref" $ pelem # outRef # inputs
```

Comment:
I posted about testing this on the plutarch discord just throw a line out.

https://discord.com/channels/922576618424250418/922577570766143498/1032400379658391592

Reply:
Plutus simple model is interesting, I kinda forgot about that.

Reply:
In follow up work, we'll need to add two levels of testing for on-chain code, with the first one being critical to verifying the logic of the protocol and second slightly less given we're already testing validators and transactions with Plutip and CTL:
- Core Protocol testing - for testing the invariant equation and the core mathematical properties of protocol actions. These tests involve simple functions that only require the values needed to produce truth values that the validator will pass on. An example of a testing a core protocol action would be the following swap test: `Given a pool with token a of A and b of B, I can swap da for db where the value of the pool is preserved.`
- Emulated Blockchain testing - for testing validators and transactions within an emulated network. An example of a test would be here would be `When minting an NFT more than once, validation fails.` It seems that [Plutus Simple Model](https://github.com/mlabs-haskell/plutus-simple-model) could be the solve here. 

@CSVdB thoughts?